src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -182,14 +185,37 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
         env.getOutputBase().getRelative(env.getRuntime().getProductName() + "-remote-logs");
     cleanAndCreateRemoteLogsDir(logDir);
 
+    if (Strings.isNullOrEmpty(remoteOptions.remoteCache) && remoteOptions.remoteCacheHeaders.size() > 0){
+      throw new AbruptExitException(
+          "--remote_cache_header can only be used if --remote_cache is set.",
+          ExitCode.COMMAND_LINE_ERROR);
+    }
+    if (Strings.isNullOrEmpty(remoteOptions.remoteExecutor) && remoteOptions.remoteCacheHeaders.size() > 0){
+      throw new AbruptExitException(
+          "--remote_exec_header can only be used if --remote_executor is set.",
+          ExitCode.COMMAND_LINE_ERROR);
+    }
     try {
       List<ClientInterceptor> interceptors = new ArrayList<>();
+      List<ClientInterceptor> execInterceptors = interceptors;
+      List<ClientInterceptor> cacheInterceptors = interceptors;
```

Comment:
This will use the same referenced list for all of the interceptors, which definitely is not what you want. You would need separate List instances, likely with interceptors populated below copied in, to effect different sets of interceptors.

Reply:
I see the reassignment below accomplishes this, but a bit awkwardly. Composition via a list view would be preferred, as jakob mentioned below, as a parameter to RemoteSpawnRunner instantiation, just so you don't have the potential for reference duplication, and don't have to rely on reference equality.

Reply:
I am not sure what you meant here. How should the headers be passed with the `RemoteSpawnRunner`? Also, that module takes care only of remote execution, doesn't it? In that case, how to pass the headers to the cache too?

Reply:
Coordinated offline. New changes look good, ty.